src/signature_accumulator.rs

```diff
@@ -0,0 +1,286 @@
+                                .is_none());
+                        }
+                        ref unexpected_msg => panic!("Unexpected message: {:?}", unexpected_msg),
+                    }
+                });
+        });
+
+        assert!(sig_accumulator.msgs.is_empty());
+        assert_eq!(sig_accumulator.sigs.len(), env.msgs_and_sigs.len());
+        sig_accumulator.sigs.values().foreach(|&(ref pub_ids_and_sigs, _)| {
+            assert_eq!(pub_ids_and_sigs.len(), env.other_ids.len())
+        });
+
+        // Add each message with the group list added - each should accumulate.
+        let mut expected_sigs_count = env.msgs_and_sigs.len();
+        assert_eq!(sig_accumulator.sigs.len(), expected_sigs_count);
+        assert!(sig_accumulator.msgs.is_empty());
+        env.msgs_and_sigs.iter().foreach(|msg_and_sigs| {
+            expected_sigs_count -= 1;
+            let mut signed_msg = msg_and_sigs.signed_msg.clone();
+            signed_msg.add_group_list(env.group_list.clone());
+            let route = rand::random();
+            let (returned_msg, returned_route) =
+                unwrap!(sig_accumulator.add_message(signed_msg.clone(), route));
+            assert_eq!(sig_accumulator.sigs.len(), expected_sigs_count);
+            assert!(sig_accumulator.msgs.is_empty());
+            assert_eq!(route, returned_route);
+            assert_eq!(signed_msg.routing_message(), returned_msg.routing_message());
+            unwrap!(returned_msg.check_integrity());
+            assert!(returned_msg.is_fully_signed());
+            env.group_list
+                .pub_ids
+                .iter()
+                .foreach(|pub_id| assert!(returned_msg.signed_by(pub_id)));
+        });
+    }
+
+    #[test]
+    fn group_src_add_signature_last() {
+        let mut sig_accumulator = SignatureAccumulator::default();
+        let env = Env::new();
+
+        // Add each message with the group list added - none should accumulate.
+        env.msgs_and_sigs.iter().enumerate().foreach(|(route, msg_and_sigs)| {
+            let mut signed_msg = msg_and_sigs.signed_msg.clone();
+            signed_msg.add_group_list(env.group_list.clone());
+            assert!(sig_accumulator.add_message(signed_msg.clone(), route as u8).is_none());
+        });
+        let mut expected_msgs_count = env.msgs_and_sigs.len();
+        assert_eq!(sig_accumulator.msgs.len(), expected_msgs_count);
+        assert!(sig_accumulator.sigs.is_empty());
+
+        // Add each message's signatures - each should accumulate once quorum has been reached.
+        env.msgs_and_sigs.iter().enumerate().foreach(|(route, msg_and_sigs)| {
+            msg_and_sigs.signature_msgs
+                .iter()
+                .zip(env.other_ids.iter())
+                .foreach(|(signature_msg, full_id)| {
+                    let result = match *signature_msg {
+                        DirectMessage::MessageSignature(ref hash, ref sig) => {
```

Comment:
Those `ref`s are unnecessary, as the values are copied below anyway. (Or is this intentional, to make the copying more explicit?)